File: app/lib/models/payment_model.dart
class Payment {
  final String id;
  final String monthlyFeeId;
  final double amount;
  final DateTime paidDate;
  final String paymentMethod; // efectivo, transferencia, cheque, tarjeta, otro
  final String? receiptNumber;
  final String? notes;
  final bool isCancelled;
  final DateTime? cancelledAt;
  final String? cancelledBy;
  final String? cancellationReason;
  final DateTime createdAt;
  final DateTime? updatedAt;

  Payment({
    required this.id,
    required this.monthlyFeeId,
    required this.amount,
    required this.paidDate,
    required this.paymentMethod,
    this.receiptNumber,
    this.notes,
    this.isCancelled = false,
    this.cancelledAt,
    this.cancelledBy,
    this.cancellationReason,
    required this.createdAt,
    this.updatedAt,
  });

  factory Payment.fromJson(Map<String, dynamic> json) {
    // Helper function to extract user ID from populated or string field
    String? extractUserId(dynamic field) {
      if (field == null) return null;
      if (field is String) return field;
      if (field is Map<String, dynamic>) return field['_id'] as String?;
      return null;
    }

    return Payment(
      id: json['_id'] ?? json['id'] ?? '',
      monthlyFeeId: json['monthlyFeeId'] ?? '',
      amount: (json['amount'] ?? 0.0).toDouble(),
      paidDate: DateTime.parse(json['paidDate'] ?? DateTime.now().toIso8601String()),
      paymentMethod: json['paymentMethod'] ?? 'efectivo',
      receiptNumber: json['receiptNumber'],
      notes: json['notes'],
      isCancelled: json['isCancelled'] ?? false,
      cancelledAt: json['cancelledAt'] != null ? DateTime.parse(json['cancelledAt']) : null,
      cancelledBy: extractUserId(json['cancelledBy']),
      cancellationReason: json['cancellationReason'],
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'monthlyFeeId': monthlyFeeId,
      'amount': amount,
      'paidDate': paidDate.toIso8601String(),
      'paymentMethod': paymentMethod,
      'receiptNumber': receiptNumber,
      'notes': notes,
      'isCancelled': isCancelled,
      'cancelledAt': cancelledAt?.toIso8601String(),
      'cancelledBy': cancelledBy,
      'cancellationReason': cancellationReason,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }

  Payment copyWith({
    String? id,
    String? monthlyFeeId,
    double? amount,
    DateTime? paidDate,
    String? paymentMethod,
    String? receiptNumber,
    String? notes,
    bool? isCancelled,
    DateTime? cancelledAt,
    String? cancelledBy,
    String? cancellationReason,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Payment(
      id: id ?? this.id,
      monthlyFeeId: monthlyFeeId ?? this.monthlyFeeId,
      amount: amount ?? this.amount,
      paidDate: paidDate ?? this.paidDate,
      paymentMethod: paymentMethod ?? this.paymentMethod,
      receiptNumber: receiptNumber ?? this.receiptNumber,
      notes: notes ?? this.notes,
      isCancelled: isCancelled ?? this.isCancelled,
      cancelledAt: cancelledAt ?? this.cancelledAt,
      cancelledBy: cancelledBy ?? this.cancelledBy,
      cancellationReason: cancellationReason ?? this.cancellationReason,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  // Getters útiles
  bool get isActive => !isCancelled;
  String get statusText => isCancelled ? 'Cancelado' : 'Activo';
}

File: api/src/models/Payment.model.js
const mongoose = require('mongoose');

const paymentSchema = new mongoose.Schema({
  monthlyFeeId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'MonthlyFee',
    required: true
  },
  amount: {
    type: Number,
    required: true,
    min: 0
  },
  paidDate: {
    type: Date,
    required: true
  },
  paymentMethod: {
    type: String,
    required: true,
    enum: ['efectivo', 'transferencia', 'cheque', 'tarjeta', 'otro'],
    default: 'efectivo'
  },
  receiptNumber: {
    type: String,
    trim: true
  },
  notes: {
    type: String,
    trim: true
  },
  isCancelled: {
    type: Boolean,
    default: false
  },
  cancelledAt: {
    type: Date
  },
  cancelledBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  cancellationReason: {
    type: String,
    trim: true
  },
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  }
}, {
  timestamps: true
});

// Índices para búsquedas frecuentes
paymentSchema.index({ monthlyFeeId: 1 });
paymentSchema.index({ paidDate: -1 });
paymentSchema.index({ isCancelled: 1 });

// Virtual para obtener el estado del pago
paymentSchema.virtual('status').get(function() {
  return this.isCancelled ? 'cancelado' : 'activo';
});

// Método estático para obtener pagos activos de una mensualidad
paymentSchema.statics.getActivePayments = function(monthlyFeeId) {
  return this.find({ 
    monthlyFeeId, 
    isCancelled: false 
  }).sort({ paidDate: -1 });
};

// Método estático para calcular total pagado de una mensualidad
paymentSchema.statics.calculateTotalPaid = function(monthlyFeeId) {
  return this.aggregate([
    {
      $match: { 
        monthlyFeeId: new mongoose.Types.ObjectId(monthlyFeeId),
        isCancelled: false 
      }
    },
    {
      $group: {
        _id: null,
        totalPaid: { $sum: '$amount' },
        paymentCount: { $sum: 1 }
      }
    }
  ]);
};

// Método para cancelar un pago
paymentSchema.methods.cancel = function(userId, reason) {
  this.isCancelled = true;
  this.cancelledAt = new Date();
  this.cancelledBy = userId;
  this.cancellationReason = reason;
  return this.save();
};

// Middleware pre-save para validaciones
paymentSchema.pre('save', function(next) {
  // Si se está cancelando, asegurar que los campos están completos
  if (this.isCancelled && !this.cancelledAt) {
    this.cancelledAt = new Date();
  }
  
  // Validar que un pago cancelado no se puede reactivar
  if (this.isModified('isCancelled') && this.isCancelled === false && this.cancelledAt) {
    return next(new Error('No se puede reactivar un pago cancelado'));
  }
  
  next();
});

module.exports = mongoose.model('Payment', paymentSchema);